scality_manila_utils/smb_helper.py

```diff
@@ -0,0 +1,361 @@
+    "TRASH-".
+
+    :param root_export: SOFS directory which holds the export points exposed
+        through manila
+    :type root_export: string (unicode)
+    :param export_name: name of export to remove
+    :type export_name: string (unicode)
+    :param exports: all the defined shares in the Samba registry
+    :type exports: dictionary
+    """
+
+    export = exports[export_name]
+    export_path = os.path.join(root_export, export_name)
+
+    # Wipe export if and only if no "external host" has access to it
+    if export['hosts allow'] not in ['', '127.0.0.1']:
+        raise ExportHasGrantsException('Unable to remove export with grants')
+
+    # We need to introduce a "variable" part (i.e a date)
+    # in case an export with the same name is deleted twice
+    tombstone = u'TRASH-{0:s}-{1:s}'.format(export_name,
+                                            time.strftime("%Y-%b-%d-%X-%Z"))
+    tombstone_path = os.path.join(root_export, tombstone)
+
+    with utils.elevated_privileges():
+        log.info("Renaming export '%s' to '%s'", export_name, tombstone)
+        try:
+            os.rename(export_path, tombstone_path)
+        except OSError as exc:
+            log.error("Unable to rename '%s' for removal : %r",
+                      export_name, exc)
+            # Two concurrent wipe_export could happen at the same time so
+            # the loser of the race could see a ENOENT.
+            if exc.errno != errno.ENOENT:
+                raise
+
+        # Persisting the parent of the moved directory is required, as
+        # it keeps track of its contents.
+        utils.fsync_path(root_export)
+
+        log.info("Deleting the export '%s' from the Samba registry",
+                 export_name)
+        process = subprocess.Popen(['net', 'conf', 'delshare', export_name],
+                                   stdout=subprocess.PIPE,
+                                   stderr=subprocess.PIPE)
+        # `subprocess.communicate` returns a byte stream
+        stdout, stderr = process.communicate()
+        stdout, stderr = stdout.decode(), stderr.decode()
+
+    if process.returncode != 0:
+        msg = ("Something went wrong while deleting the "
+               "export {0:s}: stdout={1:s}, "
+               "stderr={2:s}".format(export_name, stdout, stderr))
+        if "SBC_ERR_NO_SUCH_SERVICE" in stderr:
+            log.warning(msg)
+        else:
+            raise EnvironmentError(msg)
+
+
+def _set_hosts_allow(export_name, hosts_allow):
+    """
+    Set the `hosts_allow` parameter for a given share.
+
+    :param export_name: name of export to grant access to
+    :type export_name: string (unicode)
+    :param hosts_allow: hosts allowed on this share
+    :type hosts_allow: iterable of `str`
+    """
+
+    cmd = ['net', 'conf', 'setparm', export_name,
+           'hosts allow', ' '.join(hosts_allow)]
+    process = subprocess.Popen(cmd, stdout=subprocess.PIPE,
```

Comment:
The privilege elevation can really be put around this call. You referred to it being non-reentrant before, but I don't see how that'd be applicable. Might even have a utility function (also cfr the comment about std* logging etc) called `run_as_root` which takes `subprocess.Popen` arguments, elevates privileges, does the subprocess dance, etc.